#### Explore and cluster the neighborhoods in Toronto

In [1]:
import pandas as pd

In [5]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
dfwiki = pd. read_html(url, header = 0)

In [7]:
dftor = dfwiki[0]

In [8]:
dftor.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [9]:
dftor2 = dftor[dftor.Borough != "Not assigned"]

###### We created a new data frame by eliminating all rows with Borough "not assigned" 

In [10]:
dftor2.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

In [11]:
dftor3 = dftor2.groupby(['Postcode','Borough'], as_index = False).agg(lambda col: ', '.join(col))

In [12]:
dftor3.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

##### We now have a data frame with Neighbourhood combined into one row for the same borough. 

In [13]:
dftor3.shape

(103, 3)

##### The data has 103 rows and 3 columns

## Reading the longitude and latitude from teh csv file

In [14]:
url='http://cocl.us/Geospatial_data'

dfgeo=pd.read_csv(url)
dfgeo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
dftor3['Latitude']=dfgeo['Latitude']
dftor3['Longitude']=dfgeo['Longitude']

In [17]:
dftor3.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Exploring the data

In [20]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  49.79 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  35.22 MB/s
vincent-0.4.4- 100% |################################| Time: 0:00:00  41.02 MB/s
folium-0.5.0-p 100% |################################| Time: 0:00:00  44.13 MB/s
Libraries imported.


In [23]:
!wget -q -O 'torronto.json' dftor3

In [30]:
latitude = 43.6532
longitude = -79.3832
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(dftor3['Latitude'], dftor3['Longitude'], dftor3['Borough'], dftor3['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [31]:
CLIENT_ID = 'WPDQ1LUK52J3XPB0ZGW41CG2Q0VV5550HECBCUWIHL5APPIH' # your Foursquare ID
CLIENT_SECRET = 'LTHRTMHRQCETAISSIDJYFGG3C1DMORK5FCMAKXQMZKFQIBMF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WPDQ1LUK52J3XPB0ZGW41CG2Q0VV5550HECBCUWIHL5APPIH
CLIENT_SECRET:LTHRTMHRQCETAISSIDJYFGG3C1DMORK5FCMAKXQMZKFQIBMF


In [32]:
dftor3.loc[0, 'Neighbourhood']

'Rouge, Malvern'

In [34]:
neighbourhood_latitude = dftor3.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = dftor3.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = dftor3.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


##### top 100 venues within 1km 

In [36]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 1000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL


'https://api.foursquare.com/v2/venues/explore?&client_id=WPDQ1LUK52J3XPB0ZGW41CG2Q0VV5550HECBCUWIHL5APPIH&client_secret=LTHRTMHRQCETAISSIDJYFGG3C1DMORK5FCMAKXQMZKFQIBMF&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=1000&limit=100'

In [37]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cc08835dd579724474383e6'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4d669cba83865481c948fa53-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/spa_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1ed941735',
         'name': 'Spa',
         'pluralName': 'Spas',
         'primary': True,
         'shortName': 'Spa'}],
       'id': '4d669cba83865481c948fa53',
       'location': {'address': '8130 Sheppard Ave E',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside Ave',
        'distance': 595,
        'formattedAddress': ['8130 Sheppard Ave E (Morningside Ave)',
         'Toronto ON M1B 3W3',
         'Canada'],
        'labeledLatLngs': [{'label': 'd

In [39]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [52]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name                     categories        lat        lng
0   Images Salon & Spa                            Spa  43.802283 -79.198565
1       Caribbean Wave           Caribbean Restaurant  43.798558 -79.195777
2              Wendy's           Fast Food Restaurant  43.802008 -79.198080
3  Staples Morningside  Paper / Office Supplies Store  43.800285 -79.196607
4             Harvey's           Fast Food Restaurant  43.800106 -79.198258

In [41]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

14 venues were returned by Foursquare.


In [53]:
print(toronto_venues.shape)
toronto_venues.head()

(2238, 7)


Borough  Borough Latitude  Borough Longitude  \
0  Scarborough         43.806686         -79.194353   
1  Scarborough         43.806686         -79.194353   
2  Scarborough         43.784535         -79.160497   
3  Scarborough         43.763573         -79.188711   
4  Scarborough         43.763573         -79.188711   

                             Venue  Venue Latitude  Venue Longitude  \
0                          Wendy's       43.807448       -79.199056   
1            Interprovincial Group       43.805630       -79.200378   
2            Royal Canadian Legion       43.782533       -79.163085   
3  Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
4                G & G Electronics       43.765309       -79.191537   

         Venue Category  
0  Fast Food Restaurant  
1            Print Shop  
2                   Bar  
3           Pizza Place  
4     Electronics Store

In [54]:
toronto_venues.groupby('Borough').count()

Borough Latitude  Borough Longitude  Venue  Venue Latitude  \
Borough                                                                        
Central Toronto                118                118    118             118   
Downtown Toronto              1280               1280   1280            1280   
East Toronto                   123                123    123             123   
East York                       75                 75     75              75   
Etobicoke                       67                 67     67              67   
Mississauga                     10                 10     10              10   
North York                     237                237    237             237   
Queen's Park                    39                 39     39              39   
Scarborough                     86                 86     86              86   
West Toronto                   183                183    183             183   
York                            20                 20     20              20   

                  Venue Longitude  Venue Category  
Borough                                            
Central Toronto               118             118  
Downtown Toronto             1280            1280  
East Toronto                  123             123  
East York                      75              75  
Etobicoke                      67              67  
Mississauga                    10              10  
North York                    237             237  
Queen's Park                   39              39  
Scarborough                    86              86  
West Toronto                  183             183  
York                           20              20

#### Clustering 

In [56]:

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Borough', 
                  'Borough Latitude', 
                  'Borough Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [57]:
toronto_venues = getNearbyVenues(names=dftor3['Borough'],
                                   latitudes=dftor3['Latitude'],
                                   longitudes=dftor3['Longitude']
                                  )


Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
Scarborough
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
North York
East York
East York
East Toronto
East York
East York
East York
East Toronto
East Toronto
East Toronto
Central Toronto
Central Toronto
Central Toronto
Central Toronto
Central Toronto
Central Toronto


KeyError: 'groups'

In [58]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add borough column back to dataframe
toronto_onehot['Borough'] = toronto_venues['Borough'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Borough  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0  Scarborough                  0               0                  0        0   
1  Scarborough                  0               0                  0        0   
2  Scarborough                  0               0                  0        0   
3  Scarborough                  0               0                  0        0   
4  Scarborough                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Arcade  \
0                 0                    0             0         0       0   
1                 0                    0             0         0       0   
2                 0                    0             0         0       0   
3                 0                    0             0         0       0   
4                 0                    0             0         0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    1               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beer Bar  \
0                 0                 0                   0      0         0   
1                 0                 0                   0      0         0   
2                 0                 0                   0      0         0   
3                 0                 0                   0      0         0   
4                 0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  \
0          0         0              0                     0               0   
1          0         0              0                     0               0   
2          0         0              0                     0               0   
3          0         0              0                     0               0   
4         

In [59]:
toronto_grouped = toronto_onehot.groupby('Borough').mean().reset_index()
toronto_grouped

Borough  Accessories Store  Adult Boutique  Afghan Restaurant  \
0    Central Toronto           0.000000        0.000000           0.000000   
1   Downtown Toronto           0.000000        0.000781           0.000781   
2       East Toronto           0.000000        0.000000           0.000000   
3          East York           0.000000        0.000000           0.000000   
4          Etobicoke           0.000000        0.000000           0.000000   
5        Mississauga           0.000000        0.000000           0.000000   
6         North York           0.004219        0.000000           0.000000   
7       Queen's Park           0.000000        0.000000           0.000000   
8        Scarborough           0.000000        0.000000           0.000000   
9       West Toronto           0.000000        0.000000           0.000000   
10              York           0.000000        0.000000           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000781            0.000781      0.000781        0.001563   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.004219            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0          0.000000          0.000000             0.016949      0.000000   
1          0.001563          0.001563             0.016406      0.001563   
2          0.000000          0.000000             0.024390      0.000000   
3          0.000000          0.000000             0.000000      0.000000   
4          0.000000          0.000000             0.014925      0.000000   
5          0.000000          0.000000             0.100000      0.000000   
6          0.000000          0.000000             0.008439      0.000000   
7          0.000000          0.000000             0.000000      0.000000   
8          0.000000          0.000000             0.023256      0.000000   
9          0.000000          0.000000             0.000000      0.005464   
10         0.000000          0.000000             0.000000      0.000000   

    Aquarium    Arcade  Argentinian Restaurant  Art Gallery  Art Museum  \
0   0.000000  0.000000                0.000000     0.000000    0.000000   
1   0.003906  0.000000                0.000000     0.007812    0.001563   
2   0.000000  0.000000                0.000000     0.000000    0.000000   
3   0.000000  0.000000                0.000000     0.000000    0.000000   
4   0.000000  0.000000                0.000000     0.000000    0.000000   
5   0.000000  0.000000                0.000000     0.000000    0.000000   
6   0.000000  0.004219                0.000000     0.000000    0.000000   
7   0.000000  0.000000                0.000000     0.000000    0.000000   
8   0.000000  0.000000                0.000000     0.000000    0.000000   
9   0.000000  0.000000                0.005464     0.005464    0.000000   
10  0.000000  0.000000                0.000000     0.000000    0.000000   

    Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0              0.000000          0.000000            0.000000     0.000000   
1              0.001563          0.003906            0.000000     0.000000   
2              0.000000          0.000000            0.000000     0.000000   
3              0.000000          0.000000            0.026667     0.000000   
4              0.000000          0.000000            0.000000     0.000000   
5

In [60]:

kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Borough', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 5, 5, 2, 3, 0, 2, 4, 2, 5], dtype=int32)

In [64]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [65]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Borough']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
borough_venues_sorted = pd.DataFrame(columns=columns)
borough_venues_sorted['Borough'] = toronto_grouped['Borough']

for ind in np.arange(toronto_grouped.shape[0]):
    borough_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

borough_venues_sorted.head()

Borough 1st Most Common Venue 2nd Most Common Venue  \
0   Central Toronto           Coffee Shop        Sandwich Place   
1  Downtown Toronto           Coffee Shop                  Café   
2      East Toronto           Coffee Shop      Greek Restaurant   
3         East York           Coffee Shop                  Park   
4         Etobicoke           Pizza Place        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park           Pizza Place                  Café   
1            Restaurant                 Hotel                Bakery   
2        Ice Cream Shop    Italian Restaurant                  Café   
3              Pharmacy           Pizza Place          Burger Joint   
4              Pharmacy  Fast Food Restaurant           Coffee Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Restaurant          Dessert Shop      Sushi Restaurant   
1    Italian Restaurant                   Bar   Japanese Restaurant   
2               Brewery                Bakery                  Park   
3                  Bank   Sporting Goods Shop  Fast Food Restaurant   
4         Grocery Store                   Gym                  Café   

  9th Most Common Venue 10th Most Common Venue  
0          Burger Joint                    Pub  
1   American Restaurant     Seafood Restaurant  
2   American Restaurant            Yoga Studio  
3     Indian Restaurant         Sandwich Place  
4                  Park    Fried Chicken Joint

In [67]:
# add clustering labels
borough_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = dftor3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(borough_venues_sorted.set_index('Borough'), on='Borough')

toronto_merged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353               2  Fast Food Restaurant    Chinese Restaurant   
1 -79.160497               2  Fast Food Restaurant    Chinese Restaurant   
2 -79.188711               2  Fast Food Restaurant    Chinese Restaurant   
3 -79.216917               2  Fast Food Restaurant    Chinese Restaurant   
4 -79.239476               2  Fast Food Restaurant    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Bakery           Pizza Place           Coffee Shop   
1                Bakery           Pizza Place           Coffee Shop   
2                Bakery           Pizza Place           Coffee Shop   
3                Bakery           Pizza Place           Coffee Shop   
4                Bakery           Pizza Place           Coffee Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Breakfast Spot        Sandwich Place            Playground   
1        Breakfast Spot        Sandwich Place            Playground   
2        Breakfast Spot        Sandwich Place            Playground   
3        Breakfast Spot        Sandwich Place            Playground   
4        Breakfast Spot        Sandwich Place            Playground   

  9th Most Common Venue 10th Most Common Venue  
0                  Park               Pharmacy  
1                  Park               Pharmacy  
2                  Park               Pharmacy  
3                  Park               Pharmacy  
4                  Park               Pharmacy

In [68]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Borough'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [70]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       Scarborough               2  Fast Food Restaurant   
1       Scarborough               2  Fast Food Restaurant   
2       Scarborough               2  Fast Food Restaurant   
3       Scarborough               2  Fast Food Restaurant   
4       Scarborough               2  Fast Food Restaurant   
5       Scarborough               2  Fast Food Restaurant   
6       Scarborough               2  Fast Food Restaurant   
7       Scarborough               2  Fast Food Restaurant   
8       Scarborough               2  Fast Food Restaurant   
9       Scarborough               2  Fast Food Restaurant   
10      Scarborough               2  Fast Food Restaurant   
11      Scarborough               2  Fast Food Restaurant   
12      Scarborough               2  Fast Food Restaurant   
13      Scarborough               2  Fast Food Restaurant   
14      Scarborough               2  Fast Food Restaurant   
15      Scarborough               2  Fast Food Restaurant   
16      Scarborough               2  Fast Food Restaurant   
17       North York               2           Coffee Shop   
18       North York               2           Coffee Shop   
19       North York               2           Coffee Shop   
20       North York               2           Coffee Shop   
21       North York               2           Coffee Shop   
22       North York               2           Coffee Shop   
23       North York               2           Coffee Shop   
24       North York               2           Coffee Shop   
25       North York               2           Coffee Shop   
26       North York               2           Coffee Shop   
27       North York               2           Coffee Shop   
28       North York               2           Coffee Shop   
29       North York               2           Coffee Shop   
30       North York               2           Coffee Shop   
31       North York               2           Coffee Shop   
32       North York               2           Coffee Shop   
33       North York               2           Coffee Shop   
34       North York               2           Coffee Shop   
35        East York               2           Coffee Shop   
36        East York               2           Coffee Shop   
38        East York               2           Coffee Shop   
39        East York               2           Coffee Shop   
40        East York               2           Coffee Shop   
44  Central Toronto               2           Coffee Shop   
45  Central Toronto               2           Coffee Shop   
46  Central Toronto               2           Coffee Shop   
47  Central Toronto               2           Coffee Shop   
48  Central Toronto               2           Coffee Shop   
49  Central Toronto               2           Coffee Shop   
62       North York               2           Coffee Shop   
63  Central Toronto               2           Coffee Shop   
64  Central Toronto               2           Coffee Shop   
65  Central Toronto               2           Coffee Shop   
71       North York               2           Coffee Shop   
72       North York               2           Coffee Shop   
79       North York               2           Coffee Shop   
96       North York               2           Coffee Shop   
97       North York               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Chinese Restaurant                Bakery           Pizza Place   
1     Chinese Restaurant                Bakery           Pizza Place   
2     Chinese Restaurant                Bakery           Pizza Place   
3     Chinese Restaurant                Bakery           Pizza Place   
4     Chinese Restaurant                Bakery           Pizza Place   
5     Chinese Restaurant                Bakery           Pizza Place   
6     Chinese Restaurant                Bakery           Pizza Place   
7     Chinese Resta